### PR TITLE
`keybase deprovision`

### PR DESCRIPTION
**Pull request by oconnor663**
There's still a failing test here, but the deprovision command itself is working.

Note to self: still need to test standlaone mode.

### CODE REVIEW

**Comment by patrickxb** [comment]
Can you add some more to the test?  Try logging in after deprovision?  Check that the device keys are revoked?  Also have two users (A and B), deprovision A, make sure B still functional.

File: go/client/cmd_deprovision.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		return err
+	}
+	// XXX: This RPC deletes secret keys!
+	return loginCli.Deprovision(context.TODO(), keybase1.DeprovisionArg{
+		SessionID: 0,
+		Username:  username,
+	})
+}
+
+func NewCmdDeprovision(cl *libcmdline.CommandLine) cli.Command {
```
Comment (maxtaco) [inline go/client/cmd_deprovision.go]: pls contextify all new commands

File: go/client/cmd_deprovision.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	rpc "github.com/keybase/go-framed-msgpack-rpc"
+	"golang.org/x/net/context"
+)
+
+type CmdDeprovision struct{}
+
+func (c *CmdDeprovision) Run() (err error) {
+	protocols := []rpc.Protocol{
+		NewLogUIProtocol(),
+		NewSecretUIProtocol(G),
```
Comment (maxtaco) [inline go/client/cmd_deprovision.go]: once you've contextified the command, pls pass the local `G()` here.

File: go/client/cmd_deprovision.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	if err = RegisterProtocols(protocols); err != nil {
+		return err
+	}
+
+	username, err := c.getUsernameToDeprovision()
+	if err != nil {
+		return err
+	}
+
+	if err = GlobUI.PromptForConfirmation(getBigScaryWarning(username)); err != nil {
```
Comment (maxtaco) [inline go/client/cmd_deprovision.go]: Don't use `GlobUI` once properly contextified

File: go/client/cmd_deprovision.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		return
+	}
+
+	loginCli, err := GetLoginClient(G)
```
Comment (maxtaco) [inline go/client/cmd_deprovision.go]: ditto as above

File: go/client/cmd_deprovision.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+func (c *CmdDeprovision) GetUsage() libkb.Usage {
+	return libkb.Usage{
+		Config: true,
+		API:    true,
+	}
+}
+
+func (c *CmdDeprovision) ParseArgv(*cli.Context) error { return nil }
+
+func (c *CmdDeprovision) getUsernameToDeprovision() (string, error) {
```
Comment (maxtaco) [inline go/client/cmd_deprovision.go]: ditto below, pls don't use global `G` at all.

File: go/client/cmd_deprovision.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+func getBigScaryWarning(username string) string {
+	// TODO: Print a list of the other devices on the user's account.
+	return fmt.Sprintf(`
+%s, BE CAREFUL!  \('o')/
+
+You are about to delete this device from your account, including its secret
+keys. If you don't have any other devices, you'll lose access to your account
+and all your data!
```
Comment (maxtaco) [inline go/client/cmd_deprovision.go]: This big scary warning should say whether they have PGP keys in their secret keyring, and whether they should be nuked too.  If not, then please export via `keybase pgp export`

File: go/client/commands_common.go [hunk 7]
```diff
@@ -23,6 +23,7 @@ func GetCommands(cl *libcmdline.CommandLine, g *libkb.GlobalContext) []cli.Comma
 		NewCmdConfig(cl),
 		NewCmdCtl(cl, g),
 		NewCmdDb(cl, g),
+		NewCmdDeprovision(cl),
```
Comment (maxtaco) [inline go/client/commands_common.go]: pass `g` here

File: go/engine/deprovision.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+		return
+	}
+
+	// Then, regardless of whether the user to deprovision was logged in above,
+	// we need to nuke the local db, delete the user's secret keys, and purge
+	// the user from the local config file.
+	ctx.LogUI.Info("Clearing the local cache db...")
+	if _, err = e.G().LocalDb.Nuke(); err != nil {
+		return
+	}
```
Comment (maxtaco) [inline go/engine/deprovision.go]: there are a bunch of in-memory caches still live that should also be nuked.  See `logout` which nukes most of them.

File: go/libkb/config.go [hunk 9]
```diff
@@ -206,6 +206,27 @@ func (f *JSONConfigFile) SwitchUser(nu NormalizedUsername) error {
...
+	}
+
+	if !f.jw.AtKey("users").AtKey(nu.String()).IsNil() {
+		err := f.jw.DeleteValueAtPath("users." + nu.String())
+		if err != nil {
+			return err
+		}
+	}
+
+	return f.flush()
```
Comment (maxtaco) [inline go/libkb/config.go]: set `f.userConfigWrapper.userConfig = nil`

Comment (maxtaco) [inline go/libkb/config.go]: maybe call `setUserconfigWithLock(nil)` ?

**Comment by maxtaco** [comment]
:+1:

**Comment by oconnor663** [comment]
Thanks guys. @patrickxb yes, the test needs to expand a bunch.

File: go/client/cmd_deprovision.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	if err = RegisterProtocols(protocols); err != nil {
+		return err
+	}
+
+	username, err := c.getUsernameToDeprovision()
+	if err != nil {
+		return err
+	}
+
+	// TODO: We need to get rid of GlobUI here, but libkb.UI doesn't support PromptForConfirmation.
```
Comment (oconnor663) [inline go/client/cmd_deprovision.go]: delete this TODO

File: go/libkb/interfaces.go [hunk 11]
```diff
@@ -332,6 +333,7 @@ type UI interface {
 	GetGPGUI() GPGUI
 	GetProvisionUI(role KexRole) ProvisionUI
 	Configure() error
+	PromptForConfirmation(prompt string) error
```
Comment (oconnor663) [inline go/libkb/interfaces.go]: Let's dangle this off `TerminalUI` instead.

File: go/libkb/config.go (hunk 9 above)
Comment (oconnor663) [inline go/libkb/config.go]: It'll be cleaner to do the `SetUserConfig` here.

**Comment by maxtaco** [comment]
:clap: :clap:

**Comment by oconnor663** [comment]
@patrickxb I extended the existing test, but I'd still like to add another test that does the thing you mentioned about having multiple users log in and switching between them. Could you remind me what code we have that already tests that sort of thing?

**Pull request closed without merging**